server/matchmaker/algorithm/team_matchmaker.py

```diff
@@ -0,0 +1,286 @@
+    4. add this game to a games list
+    5. repeat 2. to 4. for every party.
+    6. you now have a list of potential games with minimal rating variation and minimal rating imbalance.
+    7. remove all games with match quality below threshold then sort by quality descending
+    8. pick the first game from the game list and remove all other games that contain the same players
+    9. repeat until the list is empty
+
+    for the future:
+    Find combination of potential games that allows for maximal number of games to launch
+    Repeat 8. with an increasing amount of top games removed beforehand to get all possible game combinations
+    If there is more than one solution, pick the one with highest total game quality
+    Optimization: sort by game quality first, abort when you find a solution with
+    the full number of theoretically possible games (floor(playersInQueue/(teamSize*2)))
+    """
+
+    def find(self, searches: Iterable[Search]) -> List[Match]:
+        self._logger.debug("========= starting matching algorithm =========")
+
+        searches = SortedList(searches, key=lambda s: s.average_rating)
+        possible_games = []
+        for index, search in enumerate(searches):
+
+            self._logger.debug("building game for %s", repr(search))
+
+            try:
+                participants = self._pick_neighboring_players(searches, index)
+                match = self.make_teams(participants)
+                game = self.assign_game_quality(match)
+                possible_games.append(game)
+            except NotEnoughPlayersException:
+                self._logger.warning("Couldn't pick enough players for a full game. Skipping this game...")
+            except UnevenTeamsException:
+                self._logger.warning("Failed to assign even teams. Skipping this game...")
+
+        self._logger.debug(
+            "got %i games\n" % len(possible_games) + "\n".join(
+                       "game: %s vs %s rating disparity: %i quality: %f" % (
+                           repr(game.match[0]),
+                           repr(game.match[1]),
+                           game.match[0].cumulated_rating - game.match[1].cumulated_rating,
+                           game.quality) for game in possible_games
+                   )
+            )
+
+        return self._pick_best_noncolliding_games(possible_games)
+
+    def _pick_neighboring_players(self, searches: List[Search], index: int) -> List[Search]:
+        """
+        Picks searches from the list starting with the search at the given index and then expanding in both directions
+        until there are enough players for a full game.
+
+        # Errors
+        May raise `NotEnoughPlayersException` if it can't find enough suitable searches to fill a game.
+        """
+        lower = iter(searches[index - 1::-1])
+        higher = iter(searches[index:])
+        i = 0
+        candidate = next(higher)
```

Comment:
This can raise `StopIteration` if higher is empty. You can use the second argument to return a default instead.
```suggestion
        candidate = next(higher, None)
```